### DISCUSSION TITLE
Transfer of a variable's derivative in MultiApp

### DISCUSSION THREAD

**Discussion by lekah** [idaholab/moose#19509] [Q&A General] [answered]
Novice user alert :smile: 

I have a master application in MOOSE that needs to access not just the (aux)variable y of a sub-application, but also the derivative of that variable, dy/dc, where c is another variable. How can I do that?
I didn't find any good way of doing that via the transfer mechanism. I can transfer the derivative separately, but then I don't know how to inform MOOSE that this thing I transferred is actually a derivative of y. 
Hope that was clear!

Any hints are appreciated, thank you so much!

**Comment by GiudGiud** [comment]
Hello

Transferring it separately is the way to go here. 
Note that auxvariables don't hold derivative information (they only hold a single field for their value, even when trying to use AD), so you ll have to code in your derivative explicitly in the kernels.

Guillaume

**Reply by lekah** [reply]
Hi Giullaume,
thanks for the answer. The variable and derivative are calculated as a material property, I just copy them into AuxVariables for the transfer. I managed to transfer them separately into the master application (again into auxvariables). What is a good way to "assemble" them again into MaterialProperty plus the derivative in the Master application?
Leonid

**Reply by GiudGiud** [reply]
You can use a MaterialRealAux to define material properties from auxiliary variables

**Reply by lekah** [reply]
Thanks for you answer, Guillaume!
In the [documentation](https://mooseframework.inl.gov/source/auxkernels/MaterialRealAux.html) , it is written that ```The MaterialRealAux AuxKernel is used to output material properties as an element-level, constant variable. ```. I am using that in the subapplication to  assign a material's property to a variable (and the derivative to another variable).
The thing missing for me is the reverse operation (in the master application): assigning an AuxVariable as a materialProperty, and another AuxVariable as the derivative of that MaterialProperty. Do you know of a way to do that?

**Reply by GiudGiud** [reply]
oh good point.
to copy an auxvariable into a material property you can use DerivativeParsedMaterial
https://mooseframework.inl.gov/source/materials/DerivativeParsedMaterial.html
but you wont be able to set the derivative, the code will try to compute it itself.

Your solution below works

**Reply by GiudGiud** [reply]
you could also have used 2 parsedmaterials to define the property and its derivative. When setting the derivative, use the name that is used to retrieve that derivative in the kernels

**Reply by lekah** [reply]
Ah, that's a good idea. Well, since I implemented a class now I will stick to that, but using 2 parsed materials and hacking the name would have been an easier way!

**Comment by lekah** [comment, marked as answer]
So, together with @GiudGiud  's answer I found a solution. 
In the subapplication, copy (field and derivative) from MaterialsProperty to AuxVariable using MaterialRealAux.
Transfer from AuxVariable in subapplication to AuxVariable in master application (field and derivatives separately).
Transfer from AuxVariable to MaterialsProperty and declare PropertyDerivative (had to write my own class for that).
